Detailed assert failure introspection for attrs and dataclasses objects

alysivji opened the pull request:
Closes https://github.com/pytest-dev/pytest/issues/3632

Adds richer equality comparison on `AssertionError` for objects creating using either attrs or dataclasses.

---

Thanks for submitting a PR, your contribution is really appreciated!

Here's a quick checklist that should be present in PRs (you can delete this text from the final description, this is just a guideline):

- [x] Create a new changelog file in the `changelog` folder, with a name like `<ISSUE NUMBER>.<TYPE>.rst`. See [changelog/README.rst](/changelog/README.rst) for details.
- [x] Target the `features` branch for new features and removals/deprecations.
- [x] Include documentation when adding new features.
- [x] Include new tests or update existing tests when applicable.

Unless your change is trivial or a small documentation fix (e.g.,  a typo or reword of a small section) please:

- [x] Add yourself to `AUTHORS` in alphabetical order;

alysivji commented on testing/test_assertion.py:
```diff
@@ -546,6 +547,146 @@ def test_mojibake(self):
+
+        @dataclass
+        class SimpleDataObject(object):
+            field_a = field()
```
I initially used type annotations here, but tests failed for 2.7 so decided to use `field` as we only care about output, not the type.

alysivji replied:
Looks like 3.7 requires type annotation. Thoughts on a good way of getting both in there?

I can edit the class dict's `__annotations__` item

RonnyPfannschmidt commented on src/_pytest/assertion/util.py:
```diff
@@ -120,6 +120,12 @@ def isdict(x):
         return isinstance(x, (set, frozenset))
 
+    def isdatacls(obj):
+        return hasattr(obj, "__dataclass_fields__")
```
please use `getattr(x, ..., None) is not None` or something like that instead, `hasattr` hides errors

RonnyPfannschmidt commented on src/_pytest/assertion/util.py:
```diff
@@ -319,6 +329,41 @@ def _compare_eq_dict(left, right, verbose=False):
 
 
+def _compare_eq_class(left, right, verbose, type=None):
+    if type == "data":
```
lets please change the magical strings to named functions that do the extraction and always require passing them

RonnyPfannschmidt commented on src/_pytest/assertion/util.py:
```diff
@@ -319,6 +329,41 @@ def _compare_eq_dict(left, right, verbose=False):
+        class_name = left.__class__.__name__
+        explanation += [("Differing attributes:")]
+        for k in diff:
+            explanation += [
```
i disagree with the chosen display style but cant provide any meaningful alternative yet - i would like to see some kind of diff style display using multiple lines

RonnyPfannschmidt commented:
@alysivji gentle ping - i'd like this to get into a mergable state - aka named functions done, and created a follow-up issue for the display of the diffs

would you like to do that, or can we pick it up?

alysivji commented:
Thanks for the ping. I will be sprinting at PyCon Canada on Monday and Tuesday next week and hope to get it done then.

I got the requested changes completed am not sure how to test Python3.7 code. I [posted a question earlier](https://github.com/pytest-dev/pytest/pull/3776#issuecomment-419742524) but looking at it now, it's not too clear in what I am asking.

I'll take some time to refamiliarize myself with the code to clarify the question I have before this weekend. Hope to hit the ground first thing Monday morning.

alysivji commented:
@RonnyPfannschmidt Got most of the changes done. Wanted to throw it into CI and get eyes on it before I get back to polishing it tomorrow.

alysivji commented:
Been waiting a few hours for the AppVeyor builds to start, but sprints are winding down and I think I'm gonna have the same error pop up in the CI. Might as well ask the question.

My commit from yesterday failed CI: `failure_demo.py` is being run using Python 2.7 and I added an example with type annotations; this results in a `SyntaxError`. How should I handle this?

[Error on AppVeyor](https://ci.appveyor.com/project/pytestbot/pytest/builds/20247278/job/et4ekm2we2oft3m5)

I also changed the formatting of the diffs. Looks a lot more readable to me.

```console
    def test_equal_data_classes():
        example1 = SimpleDataObject(1, 'b')
        example2 = SimpleDataObject(1, 'b')

        assert example1 == example2

        example3 = SimpleDataObject(1, 'xc')
>       assert example1 == example3
E       AssertionError: assert SimpleDataObj..., field_b='b') == SimpleDataObje... field_b='xc')
E         Omitting 1 identical items, use -vv to show
E         Differing attributes:
E         field_b: 'b' != 'xc'

    def test_equal_attrs():
        example1 = SimpleDataObjectAttrs(1, 'b')
        example2 = SimpleDataObjectAttrs(1, 'b')

        assert example1 == example2

        example3 = SimpleDataObjectAttrs(2, 'xc')
>       assert example1 == example3
E       AssertionError: assert SimpleDataObj..., field_b='b') == SimpleDataObje... field_b='xc')
E         Differing attributes:
E         field_a: 1 != 2
E         field_b: 'b' != 'xc'
```

with `-vv`:

```console
    def test_equal_data_classes():
        example1 = SimpleDataObject(1, 'b')
        example2 = SimpleDataObject(1, 'b')

        assert example1 == example2

        example3 = SimpleDataObject(1, 'xc')
>       assert example1 == example3
E       AssertionError: assert SimpleDataObj..., field_b='b') == SimpleDataObje... field_b='xc')
E         Matching attributes:
E         ['field_a']
E         Differing attributes:
E         field_b: 'b' != 'xc'


    def test_equal_attrs():
        example1 = SimpleDataObjectAttrs(1, 'b')
        example2 = SimpleDataObjectAttrs(1, 'b')

        assert example1 == example2

        example3 = SimpleDataObjectAttrs(2, 'xc')
>       assert example1 == example3
E       AssertionError: assert SimpleDataObj..., field_b='b') == SimpleDataObje... field_b='xc')
E         Differing attributes:
E         field_a: 1 != 2
E         field_b: 'b' != 'xc'
```

alysivji commented:
@RonnyPfannschmidt I need a hand getting this into a mergeable state

All the changes have been made, but it looks like `failure_demo.py` is being run under Py27 which does not allow for type annotations as required by the dataclass introspection this PR added.

How can I fix this?

nicoddemus commented:
@alysivji thanks for following up with this. I've committed a change so `doctesting` runs in Python 3, should be fixed now. 👍

nicoddemus commented:
@RonnyPfannschmidt anything else you would like to add here or can we merge this?
